### PR TITLE
scan utility for btc wallet

### PR DESCRIPTION
**Pull request by ZSM5J**
issue #99

### CODE REVIEW

File: cmd/scan/scan.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	_ "log"
```
Comment (gz-c) [inline cmd/scan/scan.go]: why import this way?

File: cmd/scan/scan.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+	var depTx Tx
+	depTx.BlockHeight = blockID
+	depTx.BlockHash = blockHash.String()
+	depTx.ParentHash = parentHash
+	for _, tx := range block.RawTx {
+		for i, addr := range tx.Vout {
+			depTx.TxHash = tx.Txid
+			depTx.TxHash = depTx.TxHash + ":" + strconv.Itoa(i)
+			if len(addr.ScriptPubKey.Addresses) > 0 {
+				deposits = append(deposits, Deposit{Addr: addr.ScriptPubKey.Addresses[0], Tx: depTx})
```
Comment (gz-c) [inline cmd/scan/scan.go]: Can format the `Deposit{}` with multiple lines, easier to read

File: cmd/scan/scan.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+	parentHash := block.PreviousHash
+	var deposits []Deposit
+	var depTx Tx
+	depTx.BlockHeight = blockID
+	depTx.BlockHash = blockHash.String()
+	depTx.ParentHash = parentHash
+	for _, tx := range block.RawTx {
+		for i, addr := range tx.Vout {
+			depTx.TxHash = tx.Txid
+			depTx.TxHash = depTx.TxHash + ":" + strconv.Itoa(i)
```
Comment (gz-c) [inline cmd/scan/scan.go]: `depTx.TxHash = fmt.Sprintf("%s:%d", tx.Txid, i)` is better

File: cmd/scan/scan.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+	}
+	parentHash := block.PreviousHash
+	var deposits []Deposit
+	var depTx Tx
+	depTx.BlockHeight = blockID
+	depTx.BlockHash = blockHash.String()
+	depTx.ParentHash = parentHash
+	for _, tx := range block.RawTx {
+		for i, addr := range tx.Vout {
+			depTx.TxHash = tx.Txid
```
Comment (gz-c) [inline cmd/scan/scan.go]: This assignment is redundant since you reassign the next line

File: cmd/scan/scan.go (hunk 2 above)
Comment (gz-c) [inline cmd/scan/scan.go]: Is it safe to use `Addresses[0]`? Do we know this is the correct address?

File: cmd/scan/scan.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+		}
+
+	}
+	return deposits, nil
+}
+
+func CompareAddress(addr Address, deps []Deposit) Address {
+	for _, dep := range deps {
+		if addr.Addr == dep.Addr {
+			if addr.Txs[0].BlockHeight == -1 {
```
Comment (gz-c) [inline cmd/scan/scan.go]: I don't see any code that sets `BlockHeight` to `-1` in the first place

File: cmd/scan/scan.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+				addr.Txs[0] = dep.Tx
```
Comment (gz-c) [inline cmd/scan/scan.go]: Why it always assign to `Txs[0]`?

File: cmd/scan/scan.go [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+	}
+	return false
+}
+
+func UpdateAddressInfo(addrs []Address, deps []Deposit, blockID int64) []Address {
+	branch := 0
+	for i, addr := range addrs {
+		branch = 0
+
+		if addr.MaxScanBlock == 0 && branch == 0 && blockID > 1 {
```
Comment (gz-c) [inline cmd/scan/scan.go]: you can use `if else` or a `switch` instead of `branch` flag

File: cmd/scan/scan.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+func SaveWallet(file string, addrs []Address) error {
+	wallet, err := os.Open(file)
+	defer wallet.Close()
+	if err != nil {
+		return err
+	}
+	addrsJson, _ := json.MarshalIndent(addrs, "", "    ")
+	err = ioutil.WriteFile(file, addrsJson, 0644)
```
Comment (gz-c) [inline cmd/scan/scan.go]: should return the error

File: cmd/scan/scan.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+func LoadWallet(file string) ([]Address, error) {
+	var addrs []Address
+	wallet, err := os.Open(file)
+	defer wallet.Close()
+	if err != nil {
+		return nil, err
+	}
+	jsonParser := json.NewDecoder(wallet)
+	jsonParser.Decode(&addrs)
```
Comment (gz-c) [inline cmd/scan/scan.go]: should check `Decode` error and return it

File: cmd/scan/scan.go [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+	client, err := rpcclient.New(connCfg, nil)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+func main() {
+	//btcd settings
+	btcdUser := "myuser"
```
Comment (gz-c) [inline cmd/scan/scan.go]: should be options

File: cmd/scan/scan.go (hunk 2 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: i check length upper
if len(addr.ScriptPubKey.Addresses) > 0

File: cmd/scan/scan.go (hunk 6 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: it's while address's txs is empty, when we find at least one tx func will use other if branch

File: cmd/scan/scan.go (hunk 1 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: i used log for debug and forget delete it

File: cmd/scan/scan.go (hunk 3 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: +

File: cmd/scan/scan.go (hunk 5 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: I don't how addresses will add to wallet, for now it's just putting  empty address struct in wallet.json. That's why i use blockheight=-1 as flag for zero count txs

File: cmd/scan/scan.go (hunk 10 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: done

File: cmd/scan/scan.go (hunk 2 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: done

File: cmd/scan/scan.go (hunk 4 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: reworked

File: cmd/scan/scan.go (hunk 8 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: done

File: cmd/scan/scan.go (hunk 9 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: done

File: cmd/scan/scan.go (hunk 7 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: done

File: cmd/scan/scan.go (hunk 5 above)
Comment (gz-c) [reply cmd/scan/scan.go]: This requires the user to know to put -1. Which will lead to an error.

The Txs array should be empty if there are no Txs. It should handle the initial/null case differently

Comment (ZSM5J) [reply cmd/scan/scan.go]: My way was wrong, really. I fix all and added options for adding addrs from file or from cmd.

File: cmd/scan/scan.go [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+	depTx.BlockHeight = blockID
+	depTx.BlockHash = blockHash.String()
+	depTx.ParentHash = parentHash
+	for _, tx := range block.RawTx {
+		for i, addr := range tx.Vout {
+			depTx.TxHash = fmt.Sprintf("%s:%d", tx.Txid, i)
+
+			if len(addr.ScriptPubKey.Addresses) > 0 {
+				deposits = append(deposits, Deposit{
+					Addr: addr.ScriptPubKey.Addresses[0],
```
Comment (gz-c) [inline cmd/scan/scan.go]: What does it mean if there is more than one address in here

File: cmd/scan/scan.go [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+					Tx:   depTx,
+				})
+			}
+		}
+
+	}
+	return deposits, nil
+}
+
+func CompareAddress(addr Address, deps []Deposit) Address {
```
Comment (gz-c) [inline cmd/scan/scan.go]: Rename this, "CompareAddress" does not sound like it modifies the address
Maybe change to
`func FindTxs(addr Address, deps []Deposit) []Tx`

Then later,
```
txs := FindTxs(addr, deps)
addrs[i].Txs = append(addr.Txs, txs...)
```

File: cmd/scan/scan.go [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+		}
+
+	}
+	return addrs
+}
+
+func LoadWallet(file string) ([]Address, error) {
+	var addrs []Address
+	wallet, err := os.Open(file)
+	defer wallet.Close()
```
Comment (gz-c) [inline cmd/scan/scan.go]: defer close after check error

File: cmd/scan/scan.go [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+	err = jsonParser.Decode(&addrs)
+	if err != nil {
+		return nil, err
+	}
+	return addrs, nil
+}
+
+func SaveWallet(file string, addrs []Address) error {
+	wallet, err := os.Open(file)
+	defer wallet.Close()
```
Comment (gz-c) [inline cmd/scan/scan.go]: defer close after check error

Comment (gz-c) [reply cmd/scan/scan.go]: Need to check Close() error for writable files, see
https://joeshaw.org/dont-defer-close-on-writable-files/

File: cmd/scan/scan.go [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+	user := flag.String("user", "myuser", "btcd username")
+	pass := flag.String("pass", "SomeDecentp4ssw0rd", "btcd password")
+	wallet := flag.String("wallet", "wallet.json", "wallet.json file")
+	blockN := flag.Int64("n", 0, "start blockID")
+	blockM := flag.Int64("m", 0, "finish blockID")
+	add := flag.String("add", "", "new btc addresses")
+	addFile := flag.String("add_file", "", "new btc addresses from file")
+	flag.Parse()
+
+	if len(*add) > 0 {
```
Comment (gz-c) [inline cmd/scan/scan.go]: Don't use `len` to check string empty, do `if *add != ""`

File: cmd/scan/scan.go [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+		Certificates: certs,
+	}
+	client, err := rpcclient.New(connCfg, nil)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+func main() {
```
Comment (gz-c) [inline cmd/scan/scan.go]: Should exit program with non-zero status code on error, e.g.

```
func run() error {
     // do main code
}

func main() {
    if err := run(); err != nil {
        fmt.Printf("Error: %v\n")
        os.Exit(1)
    }
}
```

File: cmd/scan/scan.go (hunk 11 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: there is one address always, can be situation when "Unable to decode output address", and this if prevent it. Look an example transaction with hash:
500f2be2ebd5193578de1e521fbd7a3d9849a8983a70a8bef89e3d0a89a9a18b

Comment (ZSM5J) [reply cmd/scan/scan.go]: I replaced it on "if addr.ScriptPubKey.Addresses != nil "

File: cmd/scan/scan.go (hunk 12 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: good

File: cmd/scan/scan.go (hunk 14 above)
Comment (ZSM5J) [reply cmd/scan/scan.go]: thanks for article

File: cmd/scan/scan.go (hunk 11 above)
Comment (gz-c) [reply cmd/scan/scan.go]: Do you have the docs for this API call/data?

Comment (ZSM5J) [reply cmd/scan/scan.go]: nope, i just test manually this moment

**Pull request merged**